### PR TITLE
Feature/multi delete

### PR DESCRIPTION
**Pull request by kuenishi**
This change reduces pain on manual testing as s3cmd changed its default behavior on recursive deletion.

c.f.: #728

### CODE REVIEW

**Comment by shino** [comment]
503 for non-existing bucket

```
s3curl.pl --id cs \
    --post \
    --contentType 'application/xml' \
    -- -s -x localhost:15018 \
    --data-binary @- \
    'http://no-such-bucket.s3.amazonaws.com/?delete' \
    <<EOF
<Delete>
  <Object>
    <Key>あ</Key>
  </Object>
</Delete>
EOF
```

```
<?xml version="1.0" encoding="UTF-8"?>
<Error>
  <Code>ServiceUnavailable</Code>
  <Message>Please reduce your request rate.</Message>
  <Resource>/no-such-bucket</Resource>
  <RequestId/>
</Error>
```

500 for non-existing key (test2 exists and owned by the user)

```
s3curl.pl --id cs \
    --post \
    --contentType 'application/xml' \
    -- -s -x localhost:15018 \
    --data-binary @- \
    'http://test2.s3.amazonaws.com/?delete' \
    <<EOF
<Delete>
  <Object>
    <Key>not-exisiting</Key>
  </Object>
</Delete>
EOF
```

```
<?xml version="1.0"?>
<html>
  <head>
    <title>500 Internal Server Error</title>
  </head>
  <body><h1>Internal Server Error</h1>The server encountered an error while processing this request</body>
</html>
```

with log

```
11:22:55.131 [error] Lager event handler error_logger_lager_h exited with reason {'EXIT',{{badmatch,["/buckets/test2/delete",{error,{error,{badmatch,{error,notfound}},[{riak_cs_wm_bucket_delete,check_permission,8,[{file,"src/riak_cs_wm_bucket_delete.erl"},{line,101}]},{riak_cs_wm_bucket_delete,'-process_post/2-fun-1-',8,[{file,"src/riak_cs_wm_bucket_delete.erl"},{line,83}]},{lists,map,2,[{file,"lists.erl"},{line,1173}]},{riak_cs_wm_bucket_delete,process_post,2,[{file,"src/riak_cs_wm_bucket_delete.erl"},{line,80}]},{webmachine_resource,resource_call,3,[{file,"src/webmac..."},...]},...]}},...]},...}}
```

File: src/riak_cs_wm_bucket_delete.erl [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        {ok, _UUIDsMarkedforDelete} ->
+            {'Deleted', [{'Key', [Key]}]};
+        Error ->
+            handle_key(RcPid, Bucket, Key, Error)
+    end.
+
+-spec parse_body(string()) -> [string()].
+parse_body(Body0) ->
+    try
+        Body = re:replace(Body0, "&quot;", "", [global, {return, list}]),
```
Comment (shino) [inline src/riak_cs_wm_bucket_delete.erl]: What is the purpose of this replace? Any trouble for keys with `"`?

File: src/riak_cs_wm_bucket_delete.erl [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        {ok, ParsedData} = riak_cs_xml:scan(Body),
+        #xmlElement{name='Delete'} = ParsedData,
```
Comment (shino) [inline src/riak_cs_wm_bucket_delete.erl]: Assertion? If so, it maybe nice to have it for avoiding invalid xml.

**Comment by shino** [comment]
Follow up of the above comment on `"` [1]

I can not delete `"` by multiple delete.

Request by s3curl:

```
s3curl.pl --id cs \
    --post \
    --contentType 'application/xml' \
    -- -s -x localhost:15018 \
    --data-binary @- \
    'http://test2.s3.amazonaws.com/?delete' \
    <<EOF
<Delete>
  <Object>
    <Key>&quot;</Key>
  </Object>
</Delete>
EOF
```

Responce includes neither `Deleted` nor `Error`

```
<?xml version="1.0" encoding="UTF-8"?>
<DeleteResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"/>
```

`"` is listed after the above request/response (2nd line of s3cmd's output).

```
$ s3cmd ls s3://test2/
                       DIR   s3://test2/</
2014-06-25 02:53      2612   s3://test2/"
```

[1] https://github.com/basho/riak_cs/pull/899/files#discussion_r14166920

**Comment by shino** [comment]
Error occurs when the name of key is `</Key>`.

```
$ s3cmd put foo s3://test2/'</Key>'
$ s3cmd -d del -r -f s3://test2/
[snip]
DEBUG: Response: {'status': 500,
 'headers': {'date': 'Wed, 25 Jun 2014 03:10:17 GMT', 'content-length': '170',
             'content-type': 'text/html', 'server': 'Riak CS'},
 'reason': 'Internal Server Error',
 'data': '<html><head><title>500 Internal Server Error</title></head><body><h1>Internal Server Error</h1>The server encountered an error while processing this request</body></html>'}
[snip]
```

(DEBUG output is formatted by hand)

By printing `ParsedData` in `riak_cs_wm_bucket_delete:parse_body/*`,
`Keys` found by xpath are two separated text nodes in xml doc tree.

https://gist.github.com/shino/e500d45d18b3d716bb7e

**Comment by shino** [comment]
Diff look good overall :+1:

**Comment by kuenishi** [comment]
All things fixed, ready again :)

**Comment by shino** [comment]
Tried to delete one object, got three errors

```
s3curl.pl --id cs \
    --post \
    --contentType 'application/xml' \
    -- -s -x localhost:15018 \
    --data-binary @- \
    'http://test2.s3.amazonaws.com/?delete' \
    <<EOF
<Delete>
  <Object>
    <Key>rebar.config</Key>
  </Object>
</Delete>
EOF
```

Response:

```
<?xml version="1.0" encoding="UTF-8"?><DeleteResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Error><Key></Key><Code>ServiceUnavailable</Code><Message>Please reduce your request rate.</Message></Error><Error><Key>rebar.config</Key><Code>ServiceUnavailable</Code><Message>Please reduce your request rate.</Message></Error><Error><Key></Key><Code>ServiceUnavailable</Code><Message>Please reduce your request rate.</Message></Error></DeleteResult>
```

Formatted resonse:

```
<?xml version="1.0" encoding="UTF-8"?>
<DeleteResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
  <Error>
    <Key/>
    <Code>ServiceUnavailable</Code>
    <Message>Please reduce your request rate.</Message>
  </Error>
  <Error>
    <Key>rebar.config</Key>
    <Code>ServiceUnavailable</Code>
    <Message>Please reduce your request rate.</Message>
  </Error>
  <Error>
    <Key/>
    <Code>ServiceUnavailable</Code>
    <Message>Please reduce your request rate.</Message>
  </Error>
</DeleteResult>
```

**Comment by shino** [comment]
It seems to conflict with #901 .

File: src/riak_cs_wm_bucket_delete.erl [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+            case riak_cs_wm_utils:check_object_authorization(Access, false, ObjectAcl,
+                                                             Policy, CanonicalId, PolicyMod,
+                                                             RcPid, BucketObj) of
+                {ok, _} -> ok;
+                {error, _} -> {error, access_denied}
+            end;
+        {error, notfound} ->
+            {error, no_such_key};
+        E ->
```
Comment (shino) [inline src/riak_cs_wm_bucket_delete.erl:133]: This `E` can be `{error,no_active_manifest}`. It's better that it results in `no_such_key`.

Comment (shino) [inline src/riak_cs_wm_bucket_delete.erl:133]: I made the same thing at https://github.com/basho/riak_cs/commit/4d23cc8?w=1#diff-42caa8070ab6ab5801654f219eed73c1R399 :see_no_evil:

File: src/riak_cs_wm_bucket_delete.erl [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+-include("riak_cs.hrl").
+-include_lib("webmachine/include/webmachine.hrl").
+-include_lib("xmerl/include/xmerl.hrl").
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-define(RIAKCPOOL, bucket_list_pool).
```
Comment (shino) [inline src/riak_cs_wm_bucket_delete.erl]: Just a curiosity, what is the reason to use `bucket_list_pool` instead of normal `request_pool`?

Comment (kuenishi) [inline src/riak_cs_wm_bucket_delete.erl]: Copy and paste from `riak_cs_wm_objects`. Mistake.

File: src/riak_cs_manifest.erl [hunk 5]
```diff
@@ -41,5 +42,13 @@ etag(?MANIFEST{content_md5={MD5, Suffix}}) ->
 etag(?MANIFEST{content_md5=MD5}) ->
     riak_cs_utils:etag_from_binary(MD5).
 
+-spec etag_no_quotes(lfs_manifest()) -> string().
```
Comment (shino) [inline src/riak_cs_manifest.erl:45]: Thanks for adding the spec :smile:

**Comment by shino** [comment]
Code look nice :+1: 

Probably another issue, better to modify mapping doc [1] 

[1] http://docs.basho.com/riakcs/latest/references/apis/storage/s3/Mapping-From-S3-API-to-Riak-CS-internal-API/#URL-Mapping

**Comment by kuenishi** [comment]
@borshop merge
yaaaaaaaaaaaaaaaaaaayyyyyyyyyyy

**Comment by michellep-basho** [comment]
Hello,

Can you please confirm if the limit is the same as for Amazon's S3 (1000 keys per request)? http://docs.aws.amazon.com/AmazonS3/latest/API/multiobjectdeleteapi.html

thanks,
Michelle

**Comment by kuenishi** [comment]
Unfortunately, no. I just missed that limit out of my mind.

**Comment by michellep-basho** [comment]
Please let me know if there's an upper limit we can include in our docs, or when one becomes available.

thanks, Michelle

**Comment by kuenishi** [comment]
@michellep I opened #912 .

**Comment by michellep-basho** [comment]
thanks, Kota.

**Pull request merged**